### PR TITLE
Refactor latex method using pylatex

### LINKED ISSUE #763: Use an external library for latex generation
### What is the expected enhancement?

The actual latex generation code that is part of the latex_drawer is tricky to maintain right now. It basically does everything by looping over the circuit and manually writing out all contents of the latex file. This includes code to track the whitespace positioning in the file (tabs, newlines etc.) and because of this the class doing the generation is quite large and hard to see how everything relates. 

There are libraries out there to try and map latex into python objects. For example pylatex: https://jeltef.github.io/PyLaTeX/current/ is a popular choice. (there are other tools out there too)

### PR DESCRIPTION
**Pull request by scottwn**
- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.

### Summary

Fixes #763 

I've removed most of the string manipulation in the `latex` method and replaced it with [pylatex](https://github.com/JelteF/PyLaTeX). I need to do more testing (and clean up all the debugging statements in the branch) but this is the initial idea.

### Details and comments

My refactor in `latex` improves this module a little bit, but it doesn't really get at the fundamental issue of maintainability. The next step would be defining pylatex subclasses for each Qcircuit command (`gate`, `cw`, etc.) and mapping directly from the circuit to those instances instead of the string representation of the commands by refactoring the `latex_array` methods. I'm not sure that would be better than the current way of doing it; we're either mapping to strings or instances and then appending a bunch of commands to `qcircuit`. I don't have a great idea for how to significantly reduce the size of the module using pylatex and am looking for suggestions, thanks!

### CODE REVIEW

**Comment by scottwn** [comment]
PR for this issue https://github.com/Qiskit/qiskit-terra/pull/4059

**Comment by ajavadia** [comment]
@scottwn are you still working on this? If/when it's ready can you remove the "WIP"

I think one obvious way to improve this file is to factor out some repetitive pieces of code into functions. For example the piece of code starting on line 443 is basically the same as the one starting on line 499. There are more examples like this.

**Comment by scottwn** [comment]
Still working! Refactoring some of the repeated code blocks, should complete this week

**Comment by scottwn** [comment]
I think these tests are failing because `pdflatex` doesn't exist in the build environment. We can add it or go back to skipping the latex tests, what do y'all think?

**Comment by 1ucian0** [comment]
Sorry this PR was neglected! If you are still interested @scottwn, can you fix the conflicts? I can help with the tests...

**Comment by scottwn** [comment]
Thanks! I'll update the branch today and then we can look at the tests.

**Comment by enavarro51** [comment]
Hi, Scott. I did a refactor of the mpl drawer a few months ago and I used this test file which sort of grew organically. It's a bit of a mess, but it did show me a lot of edge issues in the drawer. There are many things that need to be fixed in the latex drawer, and I figure if some are easy you might want to do them here, or maybe create some new issues for some of these. Just rename it .py.
[test_ctrl_gate.txt](https://github.com/Qiskit/qiskit-terra/files/5143279/test_ctrl_gate.txt)
Here's the latex output (using master. Your version may have already addressed some of these.)
testdraw7_tex
Issues to note,
- barrier and snapshot are commented out because they crash the drawer
- register names with multiple _ will also crash the drawer
- id gate should be I
- params on cu1 and zz are out of place
- inconsistent capitalizing of gate names (might be solved by using _style.disptext from qcstyle.py).
- there have been gates added that are not covered in the lists in the drawer
- reset, ctrl-swap, ctrl-rzz, and rcccx break the wires
- ctrl connections on swap and rzz go through the gates
- the ccFred... gate converts the number part of the name into e notation
- the multi-control U gates near the right side and initialize do not show parameters
- the text and mpl drawers now support labels. I assume the latex drawer should as well.

As I said, you don't need to do any of this, but just wanted a place to note some of the issues.

**Comment by mtreinish** [comment]
> Issues to note,
> 
>     * barrier and snapshot are commented out because they crash the drawer

I suspect that the version of the qcircuit latex package installed on your system is too old this is normally what causes barrier to fail for people with the latex drawer. You need the latest (released 2 years ago) version, 2.6.0, of qcircuit installed: https://ctan.org/tex-archive/graphics/qcircuit?lang=en which added the `\barrier` command.

**Comment by enavarro51** [comment]
Thanks, Matthew.

**Comment by scottwn** [comment]
Thanks guys, I need to fix some stuff with cregbundle and then I'll take a look at that list of issues

**Comment by 1ucian0** [comment]
> Thanks guys, I need to fix some stuff with cregbundle and then I'll take a look at that list of issues

Ping when ready.

**Comment by scottwn** [comment]
I've cleared most of the merge issues except the `TestLatexSourceGenerator` failures. Because the source is now being generated by PyLaTeX it doesn't match the style of the reference source files and is missing comments. Is it better to try to include the comments in the PyLaTeX source output or just replace the reference source with the PyLaTeX source as it is?

**Comment by enavarro51** [comment]
I'm not the LaTeX guy, but when we make major changes on the mpl drawer, for example the spacing between gates, that affect all the reference images, we regenerate a new set, and also give it an eyeball test.

I don't believe there's any reason to keep the format previously used for the reference latex source, so the real test is does the output produce a circuit image that's correct.

**Comment by scottwn** [comment]
I agree with that, I'm going to regenerate all the reference source files, check that they compile correctly and update the branch

**Comment by 1ucian0** [comment]
when ready, hit the "Ready for review" button!

File: requirements.txt [hunk 1]
```diff
@@ -11,3 +11,5 @@ dill>=0.3
 fastjsonschema>=2.10
 python-constraint>=1.4
 python-dateutil>=2.8.0
+pylatex>=1.3.1
+pdf2image>=1.12.1
```
Comment (1ucian0) [inline requirements.txt:15]:
```suggestion
pylatex>=1.3
pdf2image>=1.12
```

Comment (1ucian0) [inline requirements.txt:15]:
```suggestion
python-dateutil>=2.8
pylatex>=1.3
pdf2image>=1.12
```

File: setup.py [hunk 2]
```diff
@@ -36,6 +36,8 @@
     "fastjsonschema>=2.10",
     "python-constraint>=1.4",
     "python-dateutil>=2.8.0",
+    "pylatex>=1.3.1",
+    "pdf2image>=1.12.1"
```
Comment (1ucian0) [inline setup.py:40]:
```suggestion
    "python-dateutil>=2.8",
    "pylatex>=1.3",
    "pdf2image>=1.12"
```

**Comment by 1ucian0** [comment]
Some conflicting files... how far you think it might be ?

**Comment by scottwn** [comment]
The test is failing because the test environment can't compile LaTeX. Not sure what the best way is to resolve this.
```
    pylatex.errors.CompilerError: No LaTex compiler was found

Either specify a LaTex compiler or make sure you have latexmk or pdfLaTex installed.
```

**Comment by 1ucian0** [comment]
I think this PR show us that pylatex does not buy us much. Probably not worthy yet another dependency. Closing this. Thanks @scottwn for the try!

**Pull request closed without merging**